diff -rN -U0 src-3.2.0-vendor/cmake/config.cmake src-3.2.0-perturb/cmake/config.cmake
--- src-3.2.0-vendor/cmake/config.cmake
+++ src-3.2.0-perturb/cmake/config.cmake
@@ -2 +2 @@
-find_package(PythonInterp)
+find_package(PythonInterp 2)
diff -rN -U0 src-3.2.0-vendor/python/advantg/driver.py src-3.2.0-perturb/python/advantg/driver.py
--- src-3.2.0-vendor/python/advantg/driver.py
+++ src-3.2.0-perturb/python/advantg/driver.py
@@ -25,0 +26,2 @@
+from perturbation import perturbation
+
@@ -192,0 +195,3 @@
+
+    # Perturbation
+    perturbation.perturb(materials, mglib)
diff -rN -U0 src-3.2.0-vendor/python/advantg/input.py src-3.2.0-perturb/python/advantg/input.py
--- src-3.2.0-vendor/python/advantg/input.py
+++ src-3.2.0-perturb/python/advantg/input.py
@@ -14,0 +15 @@
+import perturbation
@@ -51 +52,2 @@
-    ("solver",      Module(solvers.denovo,  {"solver" : "denovo"})   )
+    ("solver",      Module(solvers.denovo,  {"solver" : "denovo"})   ),
+    ("perturbation", Module(perturbation)                            )
diff -rN -U0 src-3.2.0-vendor/python/advantg/perturbation/__init__.py src-3.2.0-perturb/python/advantg/perturbation/__init__.py
--- src-3.2.0-vendor/python/advantg/perturbation/__init__.py
+++ src-3.2.0-perturb/python/advantg/perturbation/__init__.py
@@ -0,0 +1 @@
+from options import get_ui_fields, finalize_ui_data
diff -rN -U0 src-3.2.0-vendor/python/advantg/perturbation/options.py src-3.2.0-perturb/python/advantg/perturbation/options.py
--- src-3.2.0-vendor/python/advantg/perturbation/options.py
+++ src-3.2.0-perturb/python/advantg/perturbation/options.py
@@ -0,0 +1,10 @@
+from advantg.ui import UIPath
+
+ui_fields = {}
+ui_fields['pert_temporary'] = UIPath(
+    desc='Temporary keyword',
+    isfile=True,
+    required=False)
+
+def get_ui_fields(): return ui_fields
+def finalize_ui_data(data, user_keys): return data
diff -rN -U0 src-3.2.0-vendor/python/advantg/perturbation/perturbation.py src-3.2.0-perturb/python/advantg/perturbation/perturbation.py
--- src-3.2.0-vendor/python/advantg/perturbation/perturbation.py
+++ src-3.2.0-perturb/python/advantg/perturbation/perturbation.py
@@ -0,0 +1,37 @@
+import os
+import numpy as np
+import h5py
+
+from advantg.utils import time_this
+
+def add_array_hdf5(hf, name, data):
+    hf.create_dataset(name, data=data)
+
+@time_this.with_label("Writing additional information to HDF5")
+def perturb(materials, mglib):
+    # Open HDF5 file for writing
+    if not os.path.exists('custom_output'): os.makedirs('custom_output')
+    hf = h5py.File('custom_output/xs.h5', 'w')
+
+    # List of material names
+    mat_names = np.array([mat.name for mat in materials])
+    add_array_hdf5(hf, 'mat_names', mat_names)
+
+    # Cross sections for pure materials
+    igm = mglib['igm']
+    num_mats = len(materials)
+    sigma_t = np.zeros((num_mats, igm     ))
+    sigma_s = np.zeros((num_mats, igm, igm))
+    for i, mat in enumerate(materials):
+        # Cross sections
+        xs = mglib.macro_xs[i]
+        # Total cross section
+        sigma_t[i, :] = [xs.total(x) for x in xrange(xs.igm)]
+        # Scattering matrix (g = initial group, gp = final group)
+        sigma_s[i, :, :] = [[xs.scatter(gp, g) for gp in xrange(xs.igm)]
+                            for g in xrange(xs.igm)]
+    add_array_hdf5(hf, 'sigma_t', sigma_t)
+    add_array_hdf5(hf, 'sigma_s', sigma_s)
+
+    # Close HDF5 file
+    hf.close()
diff -rN -U0 src-3.2.0-vendor/python/setup.py src-3.2.0-perturb/python/setup.py
--- src-3.2.0-vendor/python/setup.py
+++ src-3.2.0-perturb/python/setup.py
@@ -72,0 +73 @@
+            "advantg.perturbation",
